### ISSUE TITLE
Silent failure on error to create user

### ISSUE THREAD

**Issue by GregoryEAllen** [gregneagle/pycreateuserpkg#52]
The installer should report an error if:
- the username exists, but has a different uid, or
- the uid exists, but has a different username

**Comment by GregoryEAllen** [comment]
Actually, since `createuser` finds by username, the first case isn't necessarily a problem. It will update the user without changing the `uid`. The `uid` that was provided is simply ignored.

**Comment by gregneagle** [comment]
"The installer should report an error" is ambiguous.

Packages can be installed by double-clicking and using Installer.app, or via `/usr/sbin/installer`. I agree that either method should return a failure to install if creating the user fails. That is typically done by returning non-zero in the postinstall script.

How to make the error detail discoverable is a different (though related) problem -- but the answer to that is different if the package is being installed via Installer.app (the less interesting to me strategy) or via `/usr/sbin/installer`.

**Comment by gregneagle** [comment]
"Actually, since createuser finds by username, the first case isn't necessarily a problem. It will update the user without changing the uid. The uid that was provided is simply ignored."

That is an _intended_ use case -- admins can use a package to, say, create a local admin account, and then revise (and reinstall) the package to update the password for this same local admin account.

**Comment by GregoryEAllen** [comment]
Agreed, and it might be nice to know that you didn't get the `uid` that you specified.

Here are the two conditions:

```
    # uid already exists, but username does not
    if id -u $USERID >/dev/null 2>&1 && ! id -un $USERNAME >/dev/null 2>&1; then
        >&2 echo "error: uid $USERID already exists as user \"$(id -un $USERID)\""
        exit -1
    fi
    # username exists, but uid does not match
    if id -un $USERNAME >/dev/null 2>&1 && [[ $(id -u $USERNAME) != $USERID ]]; then
        >&2 echo "warning: \"$USERNAME\" has uid $(id -u $USERNAME), ignoring uid $USERID"
    fi
```

Perhaps you'd prefer to do nothing in the second case.

**Comment by GregoryEAllen** [comment]
I think echoing the error and `exit -1` is appropriate for a CLI install, but not very friendly for a GUI install. I could `osascript` a dialog if I knew how to tell which way it's being installed.

**Comment by GregoryEAllen** [comment]
From `man installer`:
```
ENVIRONMENT
     COMMAND_LINE_INSTALL  Set when performing an installation using the
                           installer command.
```

**Comment by GregoryEAllen** [comment]
Even better: `createuser` is already reporting and returning an error. It's just not getting checked.

**Comment by GregoryEAllen** [comment]
Unfortuately, `createuser` leaves a broken user when it fails. The above tests are better, unless we make changes to `createuser`.

**Comment by gregneagle** [comment]
Well, yes, a broken user account is a failure. If the user account was not broken, there would be no failure.
I think maybe you are trying to use this tool in way I didn't design it for (which is fine -- but it might mean you'll want to fork it to better meet your specific needs).
Maybe you could more fully describe your scenarios here.

I designed this tool to create a package that could be used to create a user account when a machine is first setup/configured, and to (optionally) keep the password for that same user account updated over time. Typically there are _no_ local user account present when this package is initially installed, so UID collision is not an expected issue (unless you choose UIDs < 501).

It's not really designed to create a user account on a machine that might have been deployed some time ago any might have any number of local user accounts with an unpredictable set of used UIDs. That problem is definitely solvable, but not one I needed to solve.

I might accept PRs that solve whatever issues you need solved -- but I have to make sure those changes don't negatively affect _my_ required scenarios, and that I'm comfortable supporting the changes going forward. I don't really love supporting code that I don't myself use, as I won't notice if/when it breaks, and I don't always know how to fix it.

**Comment by GregoryEAllen** [comment]
The error handing in [createuser](https://github.com/gregneagle/pycreateuserpkg/blob/c2ee28af902e6354937bed69d7aca2751c6d7554/createuser/createuser/main.m#L65) has a problem. 

In `createUserRecord` when `createRecordWithRecordType` fails a message is reported, but the erroneous record is still returned.

Then in `main`, `setAttributesForUser` gets called on the same erroneous record.

Perhaps `createUserRecord` should set the erroneous record to `nil`?

Edit: I tried it. The record still gets created, but is now empty.
